### PR TITLE
[wk-libs] increase network timeout to 30 mins, allow custom timeout

### LINKED ISSUE #569: Failed downloads with save_volume_annotation
I've found that downloading annotations using something like the code pasted below fails every now and then. The error is that "the read operation timed out." I've also pasted the full stack trace below. The error usually goes away after retrying the download a few times. I've also tried wrapping the annotation.save_volume_annotation function in a try/except, but the error still exits the running script.

Assuming I've not made a silly mistake, I was thinking a simple fix would be to pass a parameter that retries the save_volume_annotation up to some number of times.

---

The problematic code that errors periodically. I'm running this on a linux google cloud node.
```
            with wk.webknossos_context(
                    url="https://webknossos.org",
                    token=self.token):
                annotation = wk.Annotation.download(self.annotation_path)
                time_str = strftime("%Y-%m-%d_%H-%M-%S", gmtime())
                new_dataset_name = annotation.dataset_name + f"_segmented_{time_str}"  # noqa
                dataset = wk.Dataset(new_dataset_name, scale=list(self.scale))  # noqa
                annotation_layer = annotation.save_volume_annotation(dataset)  # noqa
```

The stack trace I get every now and then when the code errors out:
```
Traceback (most recent call last):
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_transports/default.py", line 61, in map_httpcore_exceptions
    yield
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_transports/default.py", line 106, in __iter__
    for part in self._httpcore_stream:
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpcore/_sync/connection_pool.py", line 57, in __iter__
    for chunk in self.stream:
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpcore/_bytestreams.py", line 56, in __iter__
    for chunk in self._iterator:
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpcore/_sync/http11.py", line 208, in _receive_response_data
    event = self._receive_event(timeout)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpcore/_sync/http11.py", line 225, in _receive_event
    data = self.socket.read(self.READ_NUM_BYTES, timeout)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpcore/_backends/sync.py", line 61, in read
    return self.sock.recv(n)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpcore/_exceptions.py", line 12, in map_exceptions
    raise to_exc(exc) from None
httpcore.ReadTimeout: The read operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 62, in _call_target
    return _target_(*args, **kwargs)
  File "/home/drew_linsley_brown_edu/contlearn/src/pl_data/dataset.py", line 148, in __init__
    volume, label = data.load()
  File "/home/drew_linsley_brown_edu/contlearn/src/pl_data/datamanager.py", line 87, in load
    annotation = wk.Annotation.download(self.annotation_path)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/webknossos/annotation/annotation.py", line 143, in download
    return download_annotation(annotation_type, annotation_id)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/webknossos/client/_download_annotation.py", line 21, in download_annotation
    response = annotation_download.sync_detailed(typ=type.value, id=id, client=client)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/webknossos/client/_generated/api/default/annotation_download.py", line 68, in sync_detailed
    response = httpx.get(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_api.py", line 189, in get
    return request(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_api.py", line 100, in request
    return client.request(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 785, in request
    return self.send(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 889, in send
    raise exc
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 880, in send
    response.read()
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_models.py", line 1463, in read
    self._content = b"".join(self.iter_bytes())
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_models.py", line 1479, in iter_bytes
    for raw_bytes in self.iter_raw():
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_models.py", line 1533, in iter_raw
    for raw_stream_bytes in self.stream:
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 127, in __iter__
    for chunk in self._stream:
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_transports/default.py", line 107, in __iter__
    yield part
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_transports/default.py", line 78, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ReadTimeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "run.py", line 202, in main
    run(cfg)
  File "run.py", line 188, in run
    trainer.fit(model=model, datamodule=datamodule)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 740, in fit
    self._call_and_handle_interrupt(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 685, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 777, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1138, in _run
    self._call_setup_hook()  # allow user to setup lightning_module in accelerator environment
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1438, in _call_setup_hook
    self.datamodule.setup(stage=fn)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/pytorch_lightning/core/datamodule.py", line 474, in wrapped_fn
    fn(*args, **kwargs)
  File "/home/drew_linsley_brown_edu/contlearn/src/pl_data/datamodule.py", line 131, in setup
    plank_train = hydra.utils.instantiate(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 180, in instantiate
    return instantiate_node(config, *args, recursive=_recursive_, convert=_convert_)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 249, in instantiate_node
    return _call_target(_target_, *args, **kwargs)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 64, in _call_target
    raise type(e)(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 62, in _call_target
    return _target_(*args, **kwargs)
  File "/home/drew_linsley_brown_edu/contlearn/src/pl_data/dataset.py", line 148, in __init__
    volume, label = data.load()
  File "/home/drew_linsley_brown_edu/contlearn/src/pl_data/datamanager.py", line 87, in load
    annotation = wk.Annotation.download(self.annotation_path)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/webknossos/annotation/annotation.py", line 143, in download
    return download_annotation(annotation_type, annotation_id)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/webknossos/client/_download_annotation.py", line 21, in download_annotation
    response = annotation_download.sync_detailed(typ=type.value, id=id, client=client)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/webknossos/client/_generated/api/default/annotation_download.py", line 68, in sync_detailed
    response = httpx.get(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_api.py", line 189, in get
    return request(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_api.py", line 100, in request
    return client.request(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 785, in request
    return self.send(
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 889, in send
    raise exc
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 880, in send
    response.read()
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_models.py", line 1463, in read
    self._content = b"".join(self.iter_bytes())
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_models.py", line 1479, in iter_bytes
    for raw_bytes in self.iter_raw():
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_models.py", line 1533, in iter_raw
    for raw_stream_bytes in self.stream:
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_client.py", line 127, in __iter__
    for chunk in self._stream:
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_transports/default.py", line 107, in __iter__
    yield part
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/drew_linsley_brown_edu/.local/lib/python3.8/site-packages/httpx/_transports/default.py", line 78, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ReadTimeout: Error instantiating 'src.pl_data.dataset.Volumetric' : The read operation timed out
```

### PR DESCRIPTION
**Pull request by jstriebel**
### Description:
- A custom network request timeout can be set using `webknossos_context(…, timeout=300)` or `export WK_TIMEOUT="300"`.
- The default network request timeout changed from ½min to 30 min.

### Issues:
- fixes #569

### Todos:
 - [x] Updated Changelog

### CODE REVIEW

**Comment by fm3** [comment]
The original error was a read timeout, not sure if this needs to be specifically configured? https://www.python-httpx.org/advanced/#fine-tuning-the-configuration

**Comment by jstriebel** [comment]
> The original error was a read timeout, not sure if this needs to be specifically configured? https://www.python-httpx.org/advanced/#fine-tuning-the-configuration

Good point, I'll try to reproduce the issue with the previous version and see which timeouts should be configure how :+1:

**Comment by jstriebel** [comment]
> The original error was a read timeout, not sure if this needs to be specifically configured? https://www.python-httpx.org/advanced/#fine-tuning-the-configuration

@fm3 I double-checked this, if a single timout is given (e.g. as a float or `httpx.Timeout(150)`), then this is used for all four different timeout-types. One can specify them additionally. E.g. 
```python
>>> httpx.Timeout(10.0, connect=60.0)
Timeout(connect=60.0, read=10.0, write=10.0, pool=10.0)
```

So this should be fine as-is.

**Comment by normanrz** [comment]
I think 2min is still pretty low. Why not go for something bigger (10min or 30min)?

**Comment by jstriebel** [comment]
> I think 2min is still pretty low. Why not go for something bigger (30min, 60min or even 120min)? These timeouts seem like nuisance when webKnossos actually needs substantial time to respond to requests.

Fine for me. I'm wondering if this might just suppress other problems for a long time (e.g. if a connection hangs), but probably this is far less a problem than timed out requests. Setting it to 30 minutes for now.

**Comment by jstriebel** [comment]
Updated it accordingly. I also sneaked in a small change to allow dataset downloads without setting the organization, defaulting to the current one from the context. @normanrz Could you have another look please?

File: webknossos/webknossos/client/context.py [hunk 1]
```diff
@@ -122,6 +125,7 @@ def _clear_all_context_caches() -> None:
 class _WebknossosContext:
     url: str = os.environ.get("WK_URL", default=DEFAULT_WEBKNOSSOS_URL)
     token: Optional[str] = os.environ.get("WK_TOKEN", default=None)
+    timeout: int = int(os.environ.get("WK_TIMEOUT", default=1800))
```
Comment (normanrz) [inline webknossos/webknossos/client/context.py:128]: Let's make a `DEFAULT_HTTP_TIMEOUT = 1800` constant

**Review by normanrz — APPROVED** [review]

**Comment by jstriebel** [comment]
@fm3 You mentioned that you're not quite happy with the name `webknossos_context`. Do you have a different proposal? I think having `context` in the name fits since it opens a context(-manager). The `webknossos` prefix is doubled a bit, one could directly use `with wk.context(…)` as well. Maybe `wk.server_context()` as an alternative? I'm quite fine with `webknossos_context`, but no strong opinion here.

**Comment by fm3** [comment]
Yeah thought about that a bit more but could not come up with a really good name either. Let’s keep it for the moment 👍

**Pull request merged**